pkg/pillar/hypervisor/pci.go

```diff
@@ -0,0 +1,200 @@
+// check availability of the VGA file in the sysfs filesystem
+func (t typeAndPCI) isVGA() bool {
+	bootVgaFile := sysfsPciDevices + t.pciLong + "/boot_vga"
+	if _, err := os.Stat(bootVgaFile); err == nil {
+		return true
+	}
+	return false
+}
+
+// read vendor ID
+func (t typeAndPCI) vid() (string, error) {
+	vendorID, err := os.ReadFile(sysfsPciDevices + t.pciLong + "/vendor")
+	if err != nil {
+		return "", err
+	}
+	return strings.TrimSpace(strings.TrimSuffix(string(vendorID), "\n")), nil
+}
+
+// isBridge checks if the given PCI device is a bridge.
+// It reads the device's class from the sysfs filesystem and returns true if the class
+// starts with "0x06", which is the PCI base class code for bridges.
+// ./class file contains 3 bytes of class code as following
+// base_class,subclass,prog-if
+// e.g. 0x06,0x04,0x00 - PCI bridge
+// PCI_BASE_CLASS_BRIDGE is defined as 0x06 in include/linux/pci_ids.h
+func (t typeAndPCI) isBridge() (bool, error) {
+	class, err := os.ReadFile(sysfsPciDevices + t.pciLong + "/class")
+
+	if err != nil {
+		logrus.Errorf("Can't read PCI device class %s: %v\n",
+			t.pciLong, err)
+		return true, err // assume it is a bridge
+	}
+
+	if strings.HasPrefix(string(class), "0x06") {
+		return true, nil
+	}
+
+	return false, nil
+}
+
+// check if the PCI device a boot_vga device
+// read the boot_vga file from the sysfs filesystem
+// and return true if the file contains "1"
+func (t typeAndPCI) isBootVGA() (bool, error) {
+	bootVGA, err := os.ReadFile(sysfsPciDevices + t.pciLong + "/boot_vga")
+
+	if err != nil {
+		logrus.Errorf("Can't read PCI device boot_vga %s: %v\n",
+			t.pciLong, err)
+		return false, err
+	}
+
+	if strings.TrimSpace(string(bootVGA)) == "1" {
+		return true, nil
+	}
+
+	return false, nil
+}
+
+// readResources reads all resources of the PCI device in a list of structs {start, end, flags}
+func (t typeAndPCI) readResources() ([]pciResource, error) {
+	var resources []pciResource
+	var resourceIndexes []int
+
+	contains := func(slice []int, item int) bool {
+		for _, s := range slice {
+			if s == item {
+				return true
+			}
+		}
+		return false
+	}
+
+	path := sysfsPciDevices + t.pciLong
+
+	// read directory for device and collect valid resource indexes
+	files, err := os.ReadDir(path)
+	if err != nil {
+		return nil, logError("Can't read PCI device directory %s: %v\n",
+			path, err)
+	}
+
+	// collect indexes of valid resources
+	for _, file := range files {
+		name := file.Name()
+		if strings.HasPrefix(name, "resource") && !strings.HasSuffix(name, "_wc") {
+			// trim prefix and convert to integer
+			resoureIndex := strings.TrimPrefix(name, "resource")
+			if resoureIndex == "" {
+				continue
+			}
+
+			index, err := strconv.Atoi(resoureIndex)
+			if err != nil {
+				return nil, logError("Can't convert PCI device resource index %s: %v\n",
+					resoureIndex, err)
+			}
+			resourceIndexes = append(resourceIndexes, index)
+		}
+	}
+
+	data, err := os.ReadFile(path + "/resource")
```

Comment:
```suggestion
	data, err := os.ReadFile(filepath.Join(path, "resource"))
```

Reply:
well, then we shuould change it everywhere :) But I agree, this is a right way to do it